src/chains/ethereum/ethereum/src/api.ts

```diff
@@ -2782,6 +2782,169 @@ export default class EthereumApi implements Api {
       overrides
     );
   }
+
+  /**
+   * Retrieves past blocks
+   *
+   * @param blockCount - Requested range of blocks. Will return less than the requested range if not all blocks are available.
+   * @param newestBlock - Highest block of the requested range.
+   * @param rewardPercentiles - (Optional) A monotonically increasing list of percentile values. For each block in the requested range, the transactions will be sorted in ascending order by effective tip per gas and the coresponding effective tip for the percentile will be determined, accounting for gas consumed.
+   *
+   * @returns The fee history for the returned block range. This can be a subsection of the requested range if not all blocks are available. Includes the oldestBlock, baseFeePerGas, gasUsedRatio, and percentile rewards.
+   * @example for: blockCount: 2, newestBlock: 0x10, rewardPercentiles: [10, 25, 50]
+   * {
+   *   oldestBlock: "0x10",
+   *   reward:[["0x59682f00","0x2a7b749c11","0x2a7b749c11"]],
+   *   baseFeePerGas:["0x40c9459e7","0x420068cc1","0x420068cc1"]
+   *   gasUsedRatio: [0.5750394293729167, 0.50]
+   * };
+   *
+   */
+  @assertArgLength(3)
+  async eth_feeHistory(
+    blockCount: QUANTITY,
+    newestBlock: QUANTITY | Ethereum.Tag,
+    rewardPercentiles: number[]
+  ): Promise<Ethereum.FeeHistoryResult> {
+    let reward = [];
+    const baseFeePerGas = [];
+    const gasUsedRatio = [];
+    const blockchain = this.#blockchain;
+    const PAD_PRECISION = 16; // Matching geth/infura decimal precision formatting
+    const PRECISION = 10 ** PAD_PRECISION;
+    const PRECISION_BIG_INT = BigInt(PRECISION);
+    const PRECISION_BIG_INT_PERCENTILE = PRECISION_BIG_INT * 100n;
+
+    let blocksToFetch = Quantity.toNumber(blockCount);
+    let newestBlockNumber = blockchain.blocks
+      .getEffectiveNumber(newestBlock)
+      .toNumber();
+    let oldestBlock;
+
+    let currentBlockNumber = newestBlockNumber;
+
+    while (blocksToFetch > 0 && currentBlockNumber >= 0) {
+      const currentBlock = await blockchain.blocks.get(
+        `0x${currentBlockNumber.toString(16)}`
+      );
+      oldestBlock = currentBlockNumber;
+
+      const gasUsed = currentBlock.header.gasUsed.toBigInt();
+      const gasLimit = currentBlock.header.gasLimit.toBigInt();
+      const baseFee = currentBlock.header.baseFeePerGas
+        ? currentBlock.header.baseFeePerGas
+        : "0x0"; // needed to match infura. Genesis blocks without EIP-1559 (e.g. forking mainnet) will be null without this.
+
+      // baseFeePerGas
+      if (currentBlockNumber === newestBlockNumber) {
+        baseFeePerGas.unshift(
+          Quantity.from(Block.calcNextBaseFee(currentBlock))
```

Comment:
:+1: @MicaiahReid from other suggested changes it became very easy to loop forward instead of going backwards. That made it easy to `calcNextBaseFee` after the loop. https://github.com/trufflesuite/ganache/pull/3351/commits/e3717579d4f2738f20293e87101d651f61acb9b0

Reply:
:+1: @jeffsmale90 https://github.com/trufflesuite/ganache/pull/3351/commits/191111f21f73833216513826d9dcb22befcf977e